### PR TITLE
Citeproc RS support

### PR DESCRIPTION
**Pull request by adomasven**
Adds experimental citeproc-rs support. Disabled behind a hidden `cite.useCiteprocRs` pref. @cormacrelf only published canary builds on NPM, so this isn't ideal in terms of NPM dependencies. But he has recently said that he'd be publishing a stable release. Also seems like my NPM got updated and this added this whole mess of a v2 package-lock file thing, which seems to be a transitory stage to a new format. Idk if we want that or not.

Related #1845

### CODE REVIEW

File: chrome/content/zotero/xpcom/style.js [hunk 1]
```diff
@@ -757,19 +761,36 @@ Zotero.Style.prototype.getCiteProc = function(locale, automaticJournalAbbreviati
...
+				}),
+				xml,
+				locale,
+				overrideLocale
+			);
+			citeproc.setOutputFormat(format);
+			citeproc.free = () => 0;
+			citeproc.opt.development_extensions.wrap_url_and_doi = true;
+			// Don't try to parse author names. We parse them in itemToCSLJSON
+			citeproc.opt.development_extensions.parse_names = false;
```
Comment (cormacrelf) [inline chrome/content/zotero/xpcom/style.js:792]: I actually have a name parser as well, do you need an option to disable it?

File: scripts/babel-worker.js [hunk 2]
```diff
@@ -79,6 +78,13 @@ async function babelWorker(ev) {
 				+ "this.singlefile.lib.getFileContent = filename => (" + JSON.stringify(webScripts) + ")[filename];\n";
 		}
 
+		// Patch citeproc_rs_wasm
+		else if (comparePaths(sourcefile, 'resource/citeproc_rs_wasm.js')) {
+			transformed = contents.replace('export class Driver', 'class Driver')
+				.replace('export default init', 'module.exports = init; module.exports.Driver = Driver')
+				.replace('input = import', '// input = import')
+				+ "\nmodule.exports = wasm_bindgen;";
+		}
```
Comment (cormacrelf) [inline scripts/babel-worker.js:87]: Is this still necessary?

Comment (adomasven) [reply scripts/babel-worker.js:87]: Nope, good spot.

File: chrome/content/zotero/xpcom/citeprocRsBridge.js [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+		// Due to Zotero being based on an old Firefox platform we are unable to use
+		// the commonJS modules generated by citeproc_rs because they use too
+		// advanced features.
+		// Cormac has decided to provide a custom build just for Zotero
+		// where citeproc_rs_wasm_include.js attaches some classes to Zotero.CiteprocRs.
+		// This works fine until we call resetDB() in the test runner (or any other time we
+		// could call ZoteroContext.reinit(). A usual require() call returns exported symbols
+		// from some Gecko script cache. Unfortunately that means that the necessary citeproc
+		// classes do not get reattached to the Zotero object and the Citeproc Wasm driver breaks
+		// when trying to create its own objects (e.g. Zotero.CiteprocRs.WasmResult).
```
[the remark this answers is not in the document]
Comment (adomasven) [reply chrome/content/zotero/xpcom/citeprocRsBridge.js:51]: You can add `await resetDB()` before https://github.com/adomasven/zotero/blob/9bc9f22fbcf1f04d04df9c0aa3313bb0bfc0aafa/test/tests/citeprocRsBridgeTest.js#L54 and then run `runtest.sh -d 5 citeprocRsBridge` which will break the test if you replace the `loadSubScript()` line above with the commented out `require('citeproc_rs_wasm_include');`

File: chrome/content/zotero/xpcom/style.js (hunk 1 above)
Comment (dstillman) [reply chrome/content/zotero/xpcom/style.js:792]: Well, I guess it's about CSL-JSON output and uses beyond our built-in citation processing, right? Does the CSL spec call for specific name-parsing behavior, or was that just something that citeproc-js implemented? I assume the goal here was just generating CSL-JSON for consumers that might not have complex name parsers built-in.

Does disabling the name parser actually make a difference if the CSL-JSON is already pre-parsed?

File: package.json [hunk 4]
```diff
@@ -15,6 +15,7 @@
   },
   "license": "",
   "dependencies": {
+    "@citeproc-rs/wasm": "^0.0.0-canary-0157c4b",
```
Comment (cormacrelf) [inline package.json:18]: I have just published [v0.1.0](https://github.com/zotero/citeproc-rs/releases/tag/wasm-v0.1.0) to [npm](https://www.npmjs.com/package/@citeproc-rs/wasm) as we speak. It doesn't include anything new, so it's just everything in master at the moment. New PRs merged will get new version bumps so you'll be able to tell the versions apart.

File: chrome/content/zotero/xpcom/citeprocRsBridge.js [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+(function() {
+
+const additionalCSLProperties = ['locator', 'label', 'suppress-author', 'author-only', 'prefix', 'suffix'];
+
+/**
+ * Bridging code which allows to use citeproc-js and citeproc-rs almost interchangeably
+ * without big alterations to existing Zotero code
+ */
+Zotero.CiteprocRs = {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/citeprocRsBridge.js:34]: Style: `CiteprocRs` → `CiteprocRS`, everywhere

Comment (adomasven) [reply chrome/content/zotero/xpcom/citeprocRsBridge.js:34]: We can't (simply) do this as the special Zotero build from CiteprocRs exports its classes onto the `Zotero.CiteprocRs` object and that's part of a [v0.1.0](https://github.com/zotero/citeproc-rs/releases/tag/wasm-v0.1.0) npm build.

Comment (dstillman) [reply chrome/content/zotero/xpcom/citeprocRsBridge.js:34]: Sad

File: chrome/content/zotero/xpcom/integration.js [hunk 6]
```diff
@@ -2038,6 +2046,9 @@ Zotero.Integration.Session.prototype.getCiteprocLists = function() {
  * Updates the list of citations to be serialized to the document
  */
 Zotero.Integration.Session.prototype._updateCitations = async function () {
+	if (Zotero.Prefs.get('cite.useCiteprocRs')) {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/integration.js:2049]: Let's at least change this to `useCiteprocRS`.

Comment (cormacrelf) [reply chrome/content/zotero/xpcom/integration.js:2049]: This doesn't really concern me, but it isn't an acronym, it's an abbreviation. JavaScript to JS is a quasi acronym, but Rust is one word that you never spell RuSt.

Comment (dstillman) [reply chrome/content/zotero/xpcom/integration.js]: OK, that's fair. We can go back to Rs.

**Comment by dstillman** [comment]
Once we update this with the line-spacing fix, I think we can go ahead and merge this.

**Comment by dstillman** [comment]
Should we merge this?

**Comment by adomasven** [comment]
I will update citeproc-rs once more with the latest changes and then I don't see a reason not to.

**Comment by adomasven** [comment]
Well, actually, we should wait for @cormacrelf to merge https://github.com/zotero/citeproc-rs/pull/142 first

**Comment by cormacrelf** [comment]
Nearly there guys, just beating the GH actions into shape. I'm merging and then cutting 0.2.0 I think.

**Comment by cormacrelf** [comment]
Alright, 0.2.0 is out on NPM!

Release notes here with the 3 action items up top. https://github.com/zotero/citeproc-rs/releases/tag/wasm-v0.2.0.

**Comment by adomasven** [comment]
Now ready to merge

**Comment by stakats** [comment]
Nice.

**Pull request merged**